spark/src/main/java/org/opensearch/sql/spark/dispatcher/AsyncQueryHandler.java

```diff
@@ -8,43 +8,95 @@
 import org.opensearch.sql.spark.dispatcher.model.DispatchQueryContext;
 import org.opensearch.sql.spark.dispatcher.model.DispatchQueryRequest;
 import org.opensearch.sql.spark.dispatcher.model.DispatchQueryResponse;
+import org.opensearch.sql.spark.execution.session.SessionId;
+import org.opensearch.sql.spark.execution.session.SessionType;
+import org.opensearch.sql.spark.execution.statement.Statement;
+import org.opensearch.sql.spark.execution.statement.StatementState;
+import org.opensearch.sql.spark.execution.statestore.StateStore;
+import org.opensearch.sql.spark.response.JobExecutionResponseReader;
 
 /** Process async query request. */
+@RequiredArgsConstructor
 public abstract class AsyncQueryHandler {
+  private final JobExecutionResponseReader jobExecutionResponseReader;
+  private final StateStore stateStore;
 
   public JSONObject getQueryResponse(AsyncQueryJobMetadata asyncQueryJobMetadata) {
-    JSONObject result = getResponseFromResultIndex(asyncQueryJobMetadata);
-    if (result.has(DATA_FIELD)) {
-      JSONObject items = result.getJSONObject(DATA_FIELD);
-
-      // If items have STATUS_FIELD, use it; otherwise, mark failed
-      String status = items.optString(STATUS_FIELD, JobRunState.FAILED.toString());
-      result.put(STATUS_FIELD, status);
-
-      // If items have ERROR_FIELD, use it; otherwise, set empty string
-      String error = items.optString(ERROR_FIELD, "");
-      result.put(ERROR_FIELD, error);
-      return result;
-    } else {
-      return getResponseFromExecutor(asyncQueryJobMetadata);
-    }
-  }
+    Statement statement = getStatementByQueryId(asyncQueryJobMetadata.getQueryId());
+    StatementState statementState = statement.getStatementState();
+    if (statementState == StatementState.SUCCESS || statementState == StatementState.FAILED) {
+      String queryId = asyncQueryJobMetadata.getQueryId().getId();
+      JSONObject result =
+          jobExecutionResponseReader.getResultWithQueryId(
+              queryId, asyncQueryJobMetadata.getResultIndex());
+      if (result.has(DATA_FIELD)) {
+        JSONObject items = result.getJSONObject(DATA_FIELD);
 
-  protected abstract JSONObject getResponseFromResultIndex(
-      AsyncQueryJobMetadata asyncQueryJobMetadata);
+        // If items have STATUS_FIELD, use it; otherwise, mark failed
+        String status = items.optString(STATUS_FIELD, JobRunState.FAILED.toString());
+        result.put(STATUS_FIELD, status);
 
-  protected abstract JSONObject getResponseFromExecutor(
-      AsyncQueryJobMetadata asyncQueryJobMetadata);
+        // If items have ERROR_FIELD, use it; otherwise, set empty string
+        String error = items.optString(ERROR_FIELD, "");
+        result.put(ERROR_FIELD, error);
+        return result;
+      }
+    }
+    JSONObject result = new JSONObject();
+    result.put(STATUS_FIELD, statementState.getState());
+    result.put(ERROR_FIELD, Optional.of(statement.getStatementModel().getError()).orElse(""));
```

Comment:
is error written to result index or statement model in request index?
Are these different cases?

Reply:
both,  Flint spark job, 
* write result, state, error, to result index.
* write state and error to request index.

@kaituo  please help confirm also.

Reply:
yes